### PR TITLE
feat(org): treat platform admins as members in internal organizations

### PR DESCRIPTION
**Pull request by tofikwest**
## What

Adds an `Organization.isInternal` flag so **platform-operated organizations** (e.g. our own) can run compliance with their platform-admin staff as **real members** — assignable, counted toward compliance progress, and included in notifications, reminders, and device rollups.

For every other organization the behavior is **unchanged**, with one deliberate exception: a global platform admin who *also* holds an `owner` member role in a non-internal org is now excluded from notifications/mentions/reminders, whereas some of those paths previously kept them. This normalizes notifications onto the same rule the assignment/approval guards already used (which never had an owner carve-out), so the participation rule is consistent everywhere. The flag defaults to `false`, and it's an **admin-only** toggle on the org admin screen.

## Why

Platform admins (global staff accounts) are intentionally excluded from an organization's day-to-day membership everywhere in the product. That's correct for the orgs they only enter to support. It's wrong for a platform-operated org, where those same accounts *are* the real members and need to run compliance like anyone else.

## How

- One shared predicate, `isOrgParticipant(userRole, { orgIsInternal })`, in `@trycompai/auth` (new `@trycompai/auth/participation` subpath so it stays dependency-light; mirrored in a dependency-free app helper for the Trigger.dev bundle).
- Replaces the scattered inline `user.role === 'admin'` participation checks across assignment/approval, compliance counting, notifications/reminders, and device queries with that single rule.
- Platform-admin **access/permission guards are intentionally untouched** — this only changes membership/participation, never privileges.
- Admin toggle: `isInternal` added to the admin org update endpoint + a "Platform settings → Internal organization" switch on the org detail screen.

## Data

- Migration `20260708120000_add_organization_is_internal` adds `isInternal BOOLEAN NOT NULL DEFAULT false`.

## Tests

- New unit tests for the predicate (jest + vitest).
- Existing task / policy / SOA / task-management / notifier suites pass unchanged, confirming behavior is preserved for orgs with the flag off.

## Rollout

Flag is off everywhere on merge. Enable per-org from the admin screen (or `PATCH /v1/admin/organizations/:id { isInternal: true }`) only for platform-operated orgs.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01UWG1YQ8KsjefFAvRV7pdtb

---
## Summary by cubic
Treat platform admins as real members in internal (platform‑operated) orgs via `Organization.isInternal`, unifying participation behind `isOrgParticipant` across API, app, and Trigger.dev. Also hardens assignment to reject non‑member IDs and fixes mention notifications to only reach the mentioned users.

- **New Features**
  - Admin toggle: “Internal organization” switch with confirmation and live refresh; `isInternal` added to the admin org update DTO and endpoint (explicit `null` now 400s).
  - Shared rule: `isOrgParticipant` in `@trycompai/auth/participation` (mirrored in the app for Trigger.dev) with helpers `isMemberOrgParticipant`, `orgParticipantMemberWhere`, and `orgParticipantMemberWhereForFlag` that return a spread‑safe AND fragment. Replaces inline checks across policies/tasks/risks/vendors/SOA, notifications/mentions, device APIs, scheduled tasks, browser automation, and the assignee picker.
  - Safety/perf: Assignee guards now reject non‑members across tasks (single/bulk), task items (create/update), policies, vendors, and risks; mention notifiers intersect (not overwrite) user filters; vendor mitigation only auto‑assigns when the author exists, is active, and is a participant; task notifier consolidates org lookup; compliance filters skip the org‑internal lookup when no platform admins are present.

- **Migration**
  - DB: adds `Organization.isInternal BOOLEAN NOT NULL DEFAULT false` (migration `20260708120000_add_organization_is_internal`).
  - Rollout: flag is off by default; enable per org via the admin UI or `PATCH /v1/admin/organizations/:id { isInternal: true }`.

<sup>Written for commit dd071b65f74ee59edb47a78660200f3771a00c71. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/trycompai/comp/pull/3378?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Comment by tofikwest** [comment]
@cubic-dev-ai review it

**Comment by tofikwest** [comment]
Reviewed the latest cubic pass (5 issues). Verdict + resolution below — fixes in `ef1a53fe5`.

**Fixed (valid):**
- **P1 · `generate-vendor-mitigation.ts` — assign author without existence guard.** Not a regression (the original `author?.user.role === 'admin' ? null : authorId` had the same behavior when the lookup missed), but a real latent bug. Hardened to fail closed: `author && isOrgParticipant(...)`, so a missing/cross-org member id can never be written.
- **P2 · `SelectAssignee.tsx` — inline rule.** Now calls the shared `isOrgParticipant` predicate instead of an inline copy, so the picker can't drift from the backend.
- **P2 · `org-participation.ts` — missing `server-only`.** Added `import 'server-only'` to match `compliance.ts` / `permissions.server.ts`.
- **P2 · duplicated rule (`org-participation-rule.ts`).** The app-local mirror is intentional (the auth index pulls better-auth, which the Trigger.dev bundler can't handle). Added a **drift-guard test** asserting the mirror stays byte-for-byte equivalent to `@trycompai/auth/participation` across every role × internal-flag combination, so CI fails if they ever diverge.

**Not a real / new issue — marking as won't-fix:**
- **P1 · `orgParticipantMemberWhere` null-role handling.** The Prisma fragment (`{ user: { role: { not: 'admin' } } }`) is **identical to `main`** — this null behavior is pre-existing, not introduced here, and the same in-memory-vs-query split existed before this PR. In practice it's moot: `User.role` is `String? @default("user")` and better-auth sets `defaultRole: 'user'`, so members effectively never have a null role. I'm intentionally **not** changing the query fragment, because doing so would alter notification/device recipient behavior on customer orgs for a case that doesn't occur. Happy to align it in a separate change if we ever allow null roles.

All source typechecks clean (only the pre-existing `CompanySubmissionWizard` errors remain) and the participation tests pass.

**Comment by tofikwest** [comment]
@cubic-dev-ai review it

**Comment by tofikwest** [comment]
Third cubic pass (3 issues) — all addressed in `77df89cdb`.

- **P1 · `orgParticipantMemberWhere` diverged from `isOrgParticipant`.** Valid consistency point (not a regression — both the owner carve-out in notifications and the strict rule in assignment came from `main`). Resolved by making the where-fragment a faithful Prisma translation of the predicate: for non-internal orgs it now excludes **only** platform admins, includes null roles explicitly (`role: { not }` skips NULL in SQL), and **drops the owner carve-out** — which the `isInternal` flag now supersedes. This also closes the earlier null-role item I'd marked won't-fix; they shared the same root. Net behavior delta is theoretical (platform-admin *owners* / null-role members in *customer* orgs effectively don't exist).
- **P2 · vendor author lookup missing `deactivated: false`.** Real edge that the internal-org path newly exposed (admins are now assignable). Added `deactivated: false` to match the other trigger tasks, so a deactivated member falls through to `null`.
- **P2 · `filterComplianceMembers` always queried `getOrgIsInternal`.** Valid perf note. Now short-circuits: the flag only matters when a platform admin is in the list, so the extra query is skipped in the common (no-admin) case. Applied to both the app and API filters.

Source typechecks clean (only the pre-existing `CompanySubmissionWizard` errors) and the participation + digest tests pass. The predicate and its Prisma translation are now a single consistent rule, guarded by the cross-package drift test added last round.

**Comment by tofikwest** [comment]
@cubic-dev-ai review it

**Comment by tofikwest** [comment]
Addressed the latest cubic pass (commit 26cecbe4c). Verdict + resolution below.

**Fixed (valid):**
- **P1 · `orgParticipantMemberWhere` overwrote a caller's `user` filter.** The helper returned a bare `{ user: {...} }`, so spreading it *after* the mention notifiers' `user: { id: { in: mentionedUserIds } }` clobbered that key — the query then matched every non-admin member instead of only the mentioned ones. The fragment is now wrapped in `AND` (via a new pure `orgParticipantMemberWhereForFlag`) so Prisma intersects it with any existing `user` condition. Added a spread-safety unit test that asserts the caller's `user` filter survives.
- **P1 · Stale `isInternal` after toggling the current org.** `OrganizationDetail` now calls `router.refresh()` when `org.id === currentOrgId`, so the server-rendered `OrgInternalProvider` (and consumers like the assignee picker) pick up the new value without a full reload.
- **P2 · No confirmation on the internal toggle.** The switch now opens an `AlertDialog` and only PATCHes after the admin confirms (both enable and disable), with copy warning it's for platform-operated orgs only.
- **P2 · Redundant query in `run-browser-automation`.** It now selects `isInternal` in the existing org lookup and uses the pure `orgParticipantMemberWhereForFlag`, dropping the extra `getOrgIsInternal` round-trip.
- **P2 · Owner carve-out inconsistency.** Cubic flagged `weekly-task-reminder`, but the same `!isOwner` carve-out was also in `task-schedule`, `policy-schedule`, and the task-item assignment notifier. Since the API where-fragment already dropped the owner carve-out (the `isInternal` flag supersedes it), I removed it from all four so every site is exactly `isOrgParticipant`. Also removed the now-dead `role` selections and stale comments.

**Extra cleanup:** fixed the admin-org test mocks that were missing the new required `isInternal` field (a typecheck break introduced when the field was added), added a `PointerEvent` shim to the vitest setup (jsdom lacks it, which base-ui's `Switch` needs), and added tests for the confirmation flow.

API participation/notifier/SOA/task-management suites pass (98); app participation + org-detail suites pass. Remaining typecheck failures are all pre-existing and in files this PR doesn't touch.

**Comment by tofikwest** [comment]
@cubic-dev-ai review it

**Comment by tofikwest** [comment]
Latest cubic pass (run `7dbbcf37`) — verdicts + resolution. Fixes in `dd071b65f`.

**Fixed (real issues):**
- **P1 × 4 · cross-org assignee** (`tasks.service`, `task-management.service`, `policies.service`, `vendors.service`): the assignee guard only threw when the org-scoped `member.findFirst` returned a member. A member id from another org — or a non-existent id — returned `null`, so the guard was skipped and the id was persisted (assignee columns don't enforce same-org membership). This gap is **pre-existing** (the prior `member?.user.role === 'admin'` check had the identical null hole), but I hardened it since the participation refactor already touches these lines. Every assignee guard now rejects a non-member with a clear 400. I applied the same fix to the sibling sites cubic didn't flag — tasks single-update, task-item update, and risks — so the whole class is consistent. Approver guards already rejected null members. Added a regression test.
- **P2 · redundant org query** (`task-notifier`): the automation-failure paths fetched the org twice (once for `isInternal` via the helper, once for `name`); consolidated into one `{ name, isInternal }` lookup + `orgParticipantMemberWhereForFlag`.

**Intentional (confirming, not a bug):**
- **P2 × 3 · owner carve-out removed** (`comment-mention`, `task-item-assignment`, `policy-schedule`): correct observation. Platform-admin **owners** of a non-internal org no longer receive these notifications. This is deliberate — the assignment/approval guards never had the owner carve-out (they always excluded all platform admins), so this makes notifications consistent with assignment, and `isInternal` is now the single mechanism for "platform admins participate here." The only affected user is a global platform admin (Comp AI staff) who is *also* an owner of a customer org, which doesn't occur in practice. The "unchanged for non-internal orgs" wording in the description is slightly too strong for that edge case; I'll tighten it. Happy to restore the carve-out if strict parity is preferred.

**Already addressed:**
- `org-participation-rule.ts` duplicate rule — covered by the cross-package drift-guard test added earlier.

Resolving these threads accordingly.

File: apps/api/src/comments/comment-mention-notifier.service.ts [hunk 1]
```diff
@@ -261,16 +262,17 @@ export class CommentMentionNotifierService {
...
       const mentionedMembers = await db.member.findMany({
         where: {
           organizationId,
           deactivated: false,
           user: { id: { in: mentionedUserIds } },
-          OR: [
-            { user: { role: { not: 'admin' } } },
-            { role: { contains: 'owner' } },
-          ],
+          ...participantWhere,
```
[the remark this answers is not in the document]
Comment (tofikwest) [reply apps/api/src/comments/comment-mention-notifier.service.ts:275]: Intentional — confirming, not fixing. The assignment/approval guards never had the owner carve-out (they always excluded all platform admins in non-internal orgs), so removing it here makes notifications consistent with assignment, and `isInternal` is now the single mechanism for platform-admin participation. The only affected user is a global platform admin who is also an owner of a customer org, which doesn't occur in practice. See the summary comment for detail; happy to restore parity if preferred.

File: apps/api/src/task-management/task-item-assignment-notifier.service.ts [hunk 2]
```diff
@@ -92,14 +92,15 @@ export class TaskItemAssignmentNotifierService {
...
 
-      // Skip notifications for platform admin members unless they are an owner
-      const isOwner = assigneeMember.role
-        ?.split(',')
-        .map((r: string) => r.trim())
-        .includes('owner');
-      if (assigneeUser.role === 'admin' && !isOwner) {
+      // Skip notifications for platform admins unless this is an internal org
+      // (where platform admins are real members) — the single participation rule.
+      if (
```
[the remark this answers is not in the document]
Comment (tofikwest) [reply apps/api/src/task-management/task-item-assignment-notifier.service.ts:97]: Intentional — confirming, not fixing. The assignment/approval guards never had the owner carve-out (they always excluded all platform admins in non-internal orgs), so removing it here makes notifications consistent with assignment, and `isInternal` is now the single mechanism for platform-admin participation. The only affected user is a global platform admin who is also an owner of a customer org, which doesn't occur in practice. See the summary comment for detail; happy to restore parity if preferred.

File: apps/app/src/trigger/tasks/task/policy-schedule.ts [hunk 3]
```diff
@@ -127,21 +126,33 @@ export const policySchedule = schedules.task({
...
-            if (!recipientsMap.has(key)) {
-              recipientsMap.set(key, {
-                email: user.email,
-                userId: user.id,
-                name: user.name ?? '',
-                policy,
-              });
-            }
+          if (!user?.email || !user.id) continue;
+          if (!isOrgParticipant(user.role, { orgIsInternal })) {
```
[the remark this answers is not in the document]
Comment (tofikwest) [reply apps/app/src/trigger/tasks/task/policy-schedule.ts:145]: Intentional — confirming, not fixing. The assignment/approval guards never had the owner carve-out (they always excluded all platform admins in non-internal orgs), so removing it here makes notifications consistent with assignment, and `isInternal` is now the single mechanism for platform-admin participation. The only affected user is a global platform admin who is also an owner of a customer org, which doesn't occur in practice. See the summary comment for detail; happy to restore parity if preferred.

File: apps/api/src/policies/policies.service.ts [hunk 4]
```diff
@@ -301,7 +300,10 @@ export class PoliciesService {
...
           include: { user: { select: { role: true } } },
         });
-        if (assignee?.user.role === 'admin') {
+        if (
+          assignee &&
+          !(await isMemberOrgParticipant(assignee.user.role, organizationId))
+        ) {
           throw new BadRequestException(
             'Cannot assign a platform admin as assignee',
           );
```
[the remark this answers is not in the document]
Comment (tofikwest) [reply apps/api/src/policies/policies.service.ts]: Fixed in `dd071b65f`. The guard now rejects a null member (cross-org or non-existent id) with a 400 before the participant check, so a foreign assignee can no longer be persisted. Applied to all assignee guards (incl. the sibling sites not flagged) for consistency; added a regression test. This gap was pre-existing but hardened here.

File: apps/api/src/tasks/tasks.service.ts [hunk 5]
```diff
@@ -472,7 +473,13 @@ export class TasksService {
           where: { id: assigneeId, organizationId },
           include: { user: { select: { role: true } } },
         });
-        if (assigneeMember?.user.role === 'admin') {
+        if (
```
[the remark this answers is not in the document]
Comment (tofikwest) [reply apps/api/src/tasks/tasks.service.ts:481]: Fixed in `dd071b65f`. The guard now rejects a null member (cross-org or non-existent id) with a 400 before the participant check, so a foreign assignee can no longer be persisted. Applied to all assignee guards (incl. the sibling sites not flagged) for consistency; added a regression test. This gap was pre-existing but hardened here.

File: apps/api/src/vendors/vendors.service.ts [hunk 6]
```diff
@@ -225,7 +226,10 @@ export class VendorsService {
       where: { id: assigneeId, organizationId },
       include: { user: { select: { role: true } } },
     });
-    if (member?.user.role === 'admin') {
+    if (
+      member &&
+      !(await isMemberOrgParticipant(member.user.role, organizationId))
+    ) {
       throw new BadRequestException(
         'Cannot assign a platform admin as assignee',
```
[the remark this answers is not in the document]
Comment (tofikwest) [reply apps/api/src/vendors/vendors.service.ts]: Fixed in `dd071b65f`. The guard now rejects a null member (cross-org or non-existent id) with a 400 before the participant check, so a foreign assignee can no longer be persisted. Applied to all assignee guards (incl. the sibling sites not flagged) for consistency; added a regression test. This gap was pre-existing but hardened here.

File: apps/api/src/task-management/task-management.service.ts [hunk 7]
```diff
@@ -304,7 +309,13 @@ export class TaskManagementService {
           where: { id: createTaskItemDto.assigneeId, organizationId },
           include: { user: { select: { role: true } } },
         });
-        if (assigneeMember?.user.role === 'admin') {
+        if (
```
[the remark this answers is not in the document]
Comment (tofikwest) [reply apps/api/src/task-management/task-management.service.ts:317]: Fixed in `dd071b65f`. The guard now rejects a null member (cross-org or non-existent id) with a 400 before the participant check, so a foreign assignee can no longer be persisted. Applied to all assignee guards (incl. the sibling sites not flagged) for consistency; added a regression test. This gap was pre-existing but hardened here.

File: apps/api/src/tasks/task-notifier.service.ts [hunk 8]
```diff
@@ -1085,6 +1086,7 @@ export class TaskNotifierService {
     } = params;
 
     try {
+      const participantWhere = await orgParticipantMemberWhere(organizationId);
```
[the remark this answers is not in the document]
Comment (tofikwest) [reply apps/api/src/tasks/task-notifier.service.ts]: Fixed in `dd071b65f` — both automation-failure paths now do a single `{ name, isInternal }` org lookup and build the participant filter with `orgParticipantMemberWhereForFlag`, dropping the redundant query.

File: apps/app/src/lib/org-participation-rule.ts [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+ * Keeping this rule free of imports lets both RSC/server code and Trigger.dev
+ * tasks share one implementation. KEEP IN SYNC with the auth package version.
+ *
+ * Platform admins (`User.role === 'admin'`) are Comp AI staff embedded in
+ * customer orgs for support; they are excluded from an org's business logic
+ * UNLESS the org is internal (platform-operated, e.g. Comp AI's own org).
+ */
+export const PLATFORM_ADMIN_ROLE = 'admin';
+
+export function isOrgParticipant(
```
[the remark this answers is not in the document]
Comment (tofikwest) [reply apps/app/src/lib/org-participation-rule.ts:17]: Addressed earlier via the cross-package drift-guard test (`org-participation-rule.test.ts`), which fails CI if the app-local mirror diverges from `@trycompai/auth/participation`. The duplication is required because the auth index pulls better-auth, which the Trigger.dev bundler can't handle.
